node/pkg/watchers/evm/blocks_by_timestamp.go

```diff
@@ -0,0 +1,149 @@
+package evm
+
+import (
+	"fmt"
+	"sort"
+	"sync"
+)
+
+const (
+	BTS_MAX_BLOCKS = 10000
+)
+
+type (
+	BlocksByTimestamp struct {
+		// cache is ordered by timestamp, blockNum. There may be multiple entries for the same timestamp, but not the same block.
+		cache Blocks
+
+		// maxCacheSize is used to trim the cache.
+		maxCacheSize int
+
+		// mutex is used to protect the cache.
+		mutex sync.Mutex
+	}
+
+	Blocks []Block
+
+	Block struct {
+		Timestamp uint64
+		BlockNum  uint64
+	}
+)
+
+// NewBlocksByTimestamp creates an empty cache of blocks by timestamp.
+func NewBlocksByTimestamp(maxCacheSize int) *BlocksByTimestamp {
+	return &BlocksByTimestamp{
+		cache:        Blocks{},
+		maxCacheSize: maxCacheSize,
+	}
+}
+
+// AddLatest adds a block to the end of the cache. This is meant to be used in the normal scenario when a new latest block is received. It will return an error if the block should not go at the end.
+func (bts *BlocksByTimestamp) AddLatest(timestamp uint64, blockNum uint64) error {
+	bts.mutex.Lock()
+	defer bts.mutex.Unlock()
+	if len(bts.cache) > 0 {
+		last := &bts.cache[len(bts.cache)-1]
+		if timestamp < last.Timestamp {
+			return fmt.Errorf("timestamp %d must be greater than or equal to the previous latest %d", timestamp, last.Timestamp)
+		}
+		if blockNum <= last.BlockNum {
+			return fmt.Errorf("block number %d must be greater than the previous latest %d", blockNum, last.BlockNum)
+		}
+	}
+
+	bts.cache = append(bts.cache, Block{Timestamp: timestamp, BlockNum: blockNum})
+
+	if len(bts.cache) > bts.maxCacheSize {
+		bts.cache = bts.cache[1:]
+	}
+	return nil
+}
+
+// AddBatch adds a batch of blocks to the cache. This is meant to be used for backfilling the cache. It makes sure there are no duplicate blocks and regenerates the cache in the correct order by timestamp.
+func (bts *BlocksByTimestamp) AddBatch(blocks Blocks) {
+	bts.mutex.Lock()
+	defer bts.mutex.Unlock()
+
+	// First build a map of all the existing blocks so we can avoid duplicates.
+	blockMap := make(map[uint64]uint64)
+	for _, block := range bts.cache {
+		blockMap[block.BlockNum] = block.Timestamp
+	}
+
+	// Now add the new blocks to the map, overwriting any duplicates. (Maybe there was a reorg. . .)
+	for _, block := range blocks {
+		blockMap[block.BlockNum] = block.Timestamp
```

Comment:
Do you need to check and have only the most recent timestamp in an overwrite?

Reply:
I would think this is intentional, I'm not sure in a reorg the timestamp would necessarily be later, but we could expect it would come in second (at least when streaming via websocket)

Reply:
Though, I am wondering if we should correct them on Safe and Finalized (to have a Safe value overwrite a Latest value and a Finalized value overwrite either)

Reply:
(now seeing that this is within Add Batch, I'm not even sure about the above comments)

Reply:
`AddBatch` will only be used for back filling with the latest data received from the RPC, so it is assumed to be correct. That's why it overwrites things.